README.rst
=======================
Password Strength Meter
=======================

A configurable, extensible password strength measuring library.


Project
=======

* Homepage: https://github.com/cadithealth/passwordmeter
* Bugs: https://github.com/cadithealth/passwordmeter/issues


TL;DR
=====

Install:

.. code-block:: bash

  $ pip install passwordmeter

Use from within an application with the default factors:

.. code-block:: python

  import passwordmeter

  strength, improvements = passwordmeter.test(sys.argv[1])

  if strength < 0.5:
    print 'Your password is too weak.'

Use on the command line:

.. code-block:: bash

  $ pwm 'password'
  Password strength: 0.132549901057 (Extremely weak)
  Possible improvements:
    - Use a good mix of numbers, letters, and symbols
    - Avoid using one of the ten thousand most common passwords
    - Use a good mix of UPPER case and lower case letters


Overview
========

The main function provided by the `passwordmeter` package is the
``Meter.test()`` method, which returns a tuple of (float, dict). The
float is the strength of the password in the range 0 to 1 (inclusive),
where 0 is extremely weak and 1 is extremely strong. The second
parameter, which may be ``None``, is a dictionary of ways the password
could be improved. The keys of the dict are general "categories" of
ways to improve the password (e.g. "length") that are fixed strings,
and the values are internationalizable strings that are human-friendly
descriptions and possibly tailored to the specific password.

A password's strength is determined by doing a weighted, skewed,
curved average of a set of "factors". The `Meter` constructor takes a
`settings` dictionary that configures, customizes, and/or supplements
the default set of factors.

The `passwordmeter.test` is a helper function that simply uses the
default settings to test the strength of a password, and is
effectively a shorthand for ``Meter().test(...)``.

For example, to use a custom selection of factors:

.. code-block:: python

  import passwordmeter

  # use only the 'length' and 'charmix' factors
  meter = passwordmeter.Meter(settings=dict(factors='length,charmix'))

  strength, improvements = meter.test('s3cr3t p4ssW0RD!')


Settings
========

The `settings` attribute to the `Meter` constructor is a dictionary
with the following keys:

* ``factors``:

  This is a comma-separated list of factors to use in calculating the
  strength of a password. Each element in the list is either the name
  of a known factor or a symbol-spec as defined by the `asset module
  <https://pypi.python.org/pypi/asset>`_. See
  ``passwordmeter.DEFAULT_FACTORS`` for the default list of factors
  (and their names).

  For example, to use only the 'length' factor and a custom factor:

  .. code-block:: python

    import passwordmeter

    class SillyFactor(passwordmeter.Factor):
      category = 'silly'
      def test(self, value, extra):
        if value == 'silly':
          return (0, 'That is a silly password!')
        return (1, None)

    meter = passwordmeter.Meter(
      settings=dict(factors=['length', SillyFactor]))

    # or, same thing, but using an asset-spec:

    meter = passwordmeter.Meter(
      settings=dict(factors='length,mypackage.SillyFactor'))

* ``factor.{NAME}.{ATTRIBUTE}``:

  Set a factor's attribute during initialization. If a setting in the
  form ``factor.{NAME}.class`` is specified for a factor not listed in
  the `factors` setting, the factor will be auto-added to the list of
  factors. This is the preferred mechanism to add a custom factor to
  the default list.

  The following attributes are "special" (all are optional):

  ===========================  ================================================
  Attribute                    Interpretation
  ===========================  ================================================
  ``factor.{NAME}.class``      Specifies the asset-spec for the factory that
                               can generate a Factor of this type.
  ``factor.{NAME}.weight``     Specifies the relative weight of this factor
                               (default: 1).
  ``factor.{NAME}.skew``       Adds the specified amount to factor score
                               (default: 0).
  ``factor.{NAME}.spread``     Multiplies the factor score by the specified
                               amount -- similar to `weight`, but is applied
                               before clipping (default: 1).
  ``factor.{NAME}.clipmin``    Force a minimum score for this factor
                               (default: 0).
  ``factor.{NAME}.clipmax``    Force a maximum score for this factor
                               (default: 1.3).
  ``factor.{NAME}.category``   Override the default improvement category.
  ===========================  ================================================

  The following example settings in an INI file will give the `length`
  factor additional weight as well as adding the "mypkg.MyFactor"
  custom factor (initialized with the parameter `msg` set to
  ``'abort'``) to the meter's list:

  .. code-block:: ini

    factor.length.weight   = 2.5
    factor.cust.class      = mypkg.MyFactor
    factor.cust.msg        = abort

* ``pessimism``:

  The password strength engine weights low scores higher than high
  scores. The degree to which the engine weights low scores is set by
  the `pessimism` setting, which defaults to 10 -- the higher, the
  more a low score will pull the average score down. For example, with
  the default pessimism of 10, the two scores 0.75 and 0.25 will be
  averaged to 0.4 (instead of the true average of 0.5).

* ``threshold``:

  Specifies the maximum score for which improvement messages should be
  returned. If not specified, all possible improvements will be
  returned, even if the relevant factor returned a perfect score (1.0
  or greater).


Custom Factors
==============

A custom factor should subclass `passwordmeter.Factor`, implement the
`test` method, and have a unique `category` (string) attribute.

The `test` method takes two parameters: the `value` to be tested, and
an opaque `extra` parameter that is supplied by the calling
application (and can be ignored if not needed). It should return a
tuple of (float, str).

The first element (float) of the return tuple must be greater or equal
to zero. Although it should generally not be greater than 1.0, a
factor *may* return a greater value: this is used to artificially
boost the strength of the total outcome relative to the other factors
if applicable. Note, however, that the Meter class will always clip
the final outcome to the inclusive range [0, 1].

The second element of the return tuple should be a string, which is a
description of how to improve the provided password. This string can
be ``None`` if no known way exists to improve this password for this
specific factor. Note that Meter class will associate this description
with the factor's category in the final outcome.


CHANGELOG.rst
=========
ChangeLog
=========


v0.1.8
======

* Removed `distribute` dependency (thanks jlec)


v0.1.7
======

* Added Python 3 support to command-line `pwm` program


v0.1.6
======

* Added Python 3 support
* First tagged release


passwordmeter/test_passwordmeter.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/10/29
# copy: (C) Copyright 2013 Cadit Health Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import unittest
import passwordmeter as pwm

class TestFactor(pwm.Factor):
  category = 'test'
  def __init__(self, prefix='test value is', *args, **kw):
    self.prefix = prefix
    super(TestFactor, self).__init__(*args, **kw)
  def test(self, value, extra):
    return (0.5, self.prefix + ': ' + value)

#------------------------------------------------------------------------------
class TestPasswordMeter(unittest.TestCase):

  #----------------------------------------------------------------------------
  def test_default(self):
    res = pwm.test('password')
    self.assertEqual(len(res), 2)
    self.assertIsInstance(res[0], float)
    self.assertIsInstance(res[1], dict)

  #----------------------------------------------------------------------------
  def test_notword(self):
    self.assertEqual(
      pwm.Meter(settings=dict(factors='notword')).test('password')[0], 0)
    self.assertEqual(
      pwm.Meter(settings=dict(factors='notword')).test('not0klsd@#$')[0], 1)

  #----------------------------------------------------------------------------
  def test_factorsAsString(self):
    self.assertEqual(
      pwm.Meter(settings=dict(
        factors='length,passwordmeter.test_passwordmeter.TestFactor')).test('short')[1],
      {'test': 'test value is: short',
       'length': 'Increase the length of the password'})

  #----------------------------------------------------------------------------
  def test_factorsAsList(self):
    self.assertEqual(
      pwm.Meter(settings=dict(
        factors=['length', TestFactor])).test('short')[1],
      {'test': 'test value is: short',
       'length': 'Increase the length of the password'})

  #----------------------------------------------------------------------------
  def test_supplementalFactor(self):
    settings = dict()
    settings['factor.test.class']  = 'passwordmeter.test_passwordmeter.TestFactor'
    settings['factor.test.prefix'] = 'test value (with prefix) is'
    res = pwm.Meter(settings=settings).test('short')
    self.assertEqual(
      sorted(res[1]),
      ['casemix', 'charmix', 'length', 'notword', 'phrase', 'test'])
    self.assertEqual(
      res[1]['test'],
      'test value (with prefix) is: short')

  #----------------------------------------------------------------------------
  def test_strength(self):
    meter = pwm.Meter()
    passwords = (
      '',
      ' ',
      'password',
      'password1',
      'pssa',
      'pssawrd',
      'pss4wr',
      'pss4wr0d',
      'p$$4wr0d!',
      'p$$4WR0d!',
      'p$4$WR0d!',
      'my voice is my p$$4WR0d!',
      'mY voiCE is my p$$4WR0d!',
      'mY voiC3 !s m-y p$$4WR0d!',
      )
    for idx, pw0 in enumerate(passwords[:-1]):
      pw1 = passwords[idx + 1]
      sc0 = meter.test(pw0)[0]
      sc1 = meter.test(pw1)[0]
      self.assertLessEqual(
        sc0, sc1,
        'expected password "%s" (%f) to be as strong or stronger than "%s" (%f)'
        % (pw1, sc1, pw0, sc0))

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------


passwordmeter/cli.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/10/26
# copy: (C) Copyright 2013 Cadit Health Inc., All Rights Reserved.
#------------------------------------------------------------------------------

from __future__ import print_function

import sys
import argparse
import getpass

from .__init__ import Meter
from .i18n import _

ratings = (
  'Infinitely weak',
  'Extremely weak',
  'Very weak',
  'Weak',
  'Moderately strong',
  'Strong',
  'Very strong',
)

#------------------------------------------------------------------------------
def main(argv=None):

  cli = argparse.ArgumentParser(
    description = _(
      'Password strength meter - gives a coefficient of how strong'
      ' a password is (0 = extremely weak, 1 = extremely strong)'
      ' and lists ways that a password can be improved.')
  )

  cli.add_argument(
    _('-v'), _('--verbose'),
    dest='verbose', default=0, action='count',
    help=_('enable verbose output to STDERR (for per-factor scoring)'))

  cli.add_argument(
    _('-i'), _('--ini'), metavar=_('FILENAME'),
    dest='inifile', default=None,
    help=_('INI settings filename'))

  cli.add_argument(
    _('-s'), _('--setting'), metavar=_('NAME=VALUE'),
    dest='settings', default=[], action='append',
    help=_('override a specific setting'))

  cli.add_argument(
    _('-m'), _('--minimum'), metavar=_('FLOAT'),
    dest='minimum', default=0.75, type=float,
    help=_('minimum password strength for %(prog)s to return'
           ' a success status (default: %(default)s)'))

  cli.add_argument(
    'password', metavar=_('PASSWORD'), nargs='?',
    help=_('password to test; if exactly "-", the password is read from'
           ' STDIN and if not specified, it will be prompted for'))

  options = cli.parse_args(args=argv)

  settings = dict()
  # todo: load ini file...
  settings.update(dict([s.split('=', 1) for s in options.settings]))

  if options.password == '-':
    options.password = sys.stdin.read()
  elif options.password is None:
    options.password = getpass.getpass()

  if options.verbose:
    # todo: use `logging`?...
    class mylogger(object):
      def debug(self, msg, *args, **kw):
        if args:
          msg = msg % args
        sys.stderr.write(msg + '\n')
    settings['logger'] = mylogger()

  meter   = Meter(settings=settings)
  result  = meter.test(options.password)

  print('Password strength: {} ({})'.format(
    result[0],
    _(ratings[min(len(ratings) - 1, int(result[0] * len(ratings)))])
  ))

  if result[1]:
    print('Possible improvements:')
    for item in result[1].values():
      print('  -',item)

  if result[0] >= options.minimum:
    return 0
  return 100 + int(round(result[0] * 100))

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------


passwordmeter/__init__.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <grabner@cadit.com>
# date: 2013/10/26
# copy: (C) Copyright 2013 Cadit Health Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import re
import asset
import pkg_resources

from .i18n import _

#------------------------------------------------------------------------------
DEFAULT_INFLECT = 0.75
DEFAULT_WEIGHT  = 1.0
DEFAULT_CLIPMIN = 0
DEFAULT_CLIPMAX = 1.3
DEFAULT_SKEW    = 0
DEFAULT_SPREAD  = 1.0
DEFAULT_FACTORS = 'length,charmix,variety,casemix,notword,phrase'

#------------------------------------------------------------------------------
common10k = frozenset(
  asset.load('passwordmeter:res/common.txt').read().decode().split('\n'))

#------------------------------------------------------------------------------
def asym(value, target, switch=DEFAULT_INFLECT):
  if value >= target:
    return 1 - ( ( 1 - switch ) * target / value )
  return switch * value / target

#------------------------------------------------------------------------------
def curve(value, offset=0.05):
  if value < 0:
    value = 0
  return 1.0 / ( offset + value )

#------------------------------------------------------------------------------
def curveavg(values):
  score  = 0
  weight = 0
  for v in values:
    w = curve(v)
    score  += v * w
    weight += w
  return score / weight

#------------------------------------------------------------------------------
class Factor(object):
  def __init__(self, weight=DEFAULT_WEIGHT,
               clipmin=DEFAULT_CLIPMIN, clipmax=DEFAULT_CLIPMAX,
               skew=DEFAULT_SKEW, spread=DEFAULT_SPREAD,
               category=None,
               ):
    self.weight     = float(weight)
    self.clipmin    = float(clipmin)
    self.clipmax    = float(clipmax)
    self.skew       = float(skew)
    self.spread     = float(spread)
    if category is not None:
      self.category = category
  def test(self, value, extra):
    raise NotImplementedError()
  def adjust(self, value):
    value = self.skew + self.spread * value
    if value < self.clipmin:
      value = self.clipmin
    if value > self.clipmax:
      value = self.clipmax
    return value, self.weight

#------------------------------------------------------------------------------
class LengthFactor(Factor):
  category = 'length'
  def __init__(self, target=8, *args, **kw):
    super(LengthFactor, self).__init__(*args, **kw)
    self.target = int(target)
  def test(self, value, extra):
    value = len(value)
    return (asym(value, self.target), _('Increase the length of the password'))

#------------------------------------------------------------------------------
class CharmixFactor(Factor):
  category = 'charmix'
  matchers = (
    re.compile('[a-zA-Z]'),     # alpha / TODO: what about non-ascii letters?
    re.compile('[0-9]'),        # numeric
    re.compile('[^a-zA-Z0-9]'), # symbols
    )
  message = 'Use a good mix of numbers, letters, and symbols'
  def test(self, value, extra):
    if not self.matchers:
      return 1.0
    scores    = [len(matcher.findall(value)) for matcher in self.matchers]
    target    = max(1, 0.25 * sum(scores) / len(scores))
    scores    = [asym(score, target) * 1 / DEFAULT_INFLECT for score in scores]
    return (curveavg(scores), _(self.message))

#------------------------------------------------------------------------------
class CasemixFactor(CharmixFactor):
  category = 'casemix'
  matchers = (
    re.compile('[a-z]'),
    re.compile('[A-Z]'),
    )
  message = 'Use a good mix of UPPER case and lower case letters'

#------------------------------------------------------------------------------
class VarietyFactor(Factor):
  category = 'variety'
  def test(self, value, extra):
    diff = 1.0
    same = 0.0
    for idx in range(1, len(value)):
      if value[idx] == value[idx - 1]:
        same += 1
      else:
        diff += 1
    score = curveavg([
      len(set(value)) / max(1, float(len(value))),
      diff / (diff + same ** 1.5)])
    return (score, _('Minimize character duplicates and repetitions'))

#------------------------------------------------------------------------------
class NotWordFactor(Factor):
  category = 'notword'
  def test(self, value, extra):
    if value.lower() in common10k:
      return (0, _('Avoid using one of the ten thousand most common passwords'))
    # TODO: check against dictionary words too... maybe use ispell?
    #         http://code.activestate.com/recipes/117221-spell-checking/
    # todo: check for 'l33t-speak'...
    return (1.0, None)

#------------------------------------------------------------------------------
class PhraseFactor(Factor):
  category = 'phrase'
  base = 0.65
  message = 'Passphrases (e.g. an obfuscated sentence) are better than passwords'
  def test(self, value, extra):
    # todo: this might be improved... for example, i should:
    #   - value words of different lengths (large distribution) -- check stddev
    #   - check the likelihood that the elements are in fact words
    wlens  = [len(s) for s in value.split()]
    if not wlens:
      return (0, _(self.message))
    score  = asym(len(wlens), 4, switch=self.base)
    spread = max(wlens) - min(wlens)
    if spread <= 0:
      score = self.base
    else:
      score *= asym(spread, 4, switch=self.base)
      score = self.base * ( 1 + score )
    return (score, _(self.message))

#------------------------------------------------------------------------------
class Meter(object):

  #----------------------------------------------------------------------------
  def __init__(self, settings=None):
    if settings is None:
      settings = dict()
    self.threshold = float(settings.get('threshold', DEFAULT_INFLECT))
    factors = settings.get('factors') or DEFAULT_FACTORS
    if asset.isstr(factors):
      factors = [s.strip() for s in factors.split(',')]
    for key in settings.keys():
      key = key.split('.')
      if len(key) >= 3 \
          and key[0] == 'factor' and key[2] == 'class' \
          and key[1] not in factors:
        factors.append(key[1])
    self.factors   = [self._load(factor, settings) for factor in factors]
    self.logger    = asset.symbol(settings.get('logger'))
    self.pessimism = 1.0 / float(settings.get('pessimism', 10))

  #----------------------------------------------------------------------------
  def _load(self, factor, settings):
    predef = {
      'length'   : LengthFactor,
      'charmix'  : CharmixFactor,
      'casemix'  : CasemixFactor,
      'variety'  : VarietyFactor,
      'notword'  : NotWordFactor,
      'phrase'   : PhraseFactor,
      }
    params = dict()
    if asset.isstr(factor):
      for key, value in settings.items():
        if key.startswith('factor.' + factor + '.'):
          params[key[len(factor) + 8:]] = value
    factor = params.pop('class', factor)
    if factor in predef:
      return predef[factor](**params)
    return asset.symbol(factor)(**params)

  #----------------------------------------------------------------------------
  def test(self, value, extra=None):
    '''
    Tests the supplied `value` for general characteristics of a good
    password. Returns a tuple of (float, dict), where float is a
    quotient in the range [0, 1] representing very weak to very strong
    passwords. dict, which may be None, is a list of known ways that
    the password can be improved, where the key is a fixed value (from
    the list below or from the settings in the constructor) and the
    value is a human-readable (potentially internationalized) string.

    The optional parameter `extra` can be any kind of object that is
    passed through unaltered to all factors contributing to this
    meter. This is primarily for custom :class:`Factor`
    implementations that need extra information -- it is not used by
    any of the standard factors.
    '''
    score  = 0
    weight = 0
    morelist = []
    for factor in self.factors:
      result   = factor.test(value, extra)
      scr, wgt = factor.adjust(result[0])
      crv      = curve(scr, offset=self.pessimism)
      if self.logger:
        self.logger.debug(
          'factor %s: score=%f, base-weight=%f, curve-weight=%f',
          factor.category, scr, wgt, crv)
      wgt    *= crv
      score  += scr * wgt
      weight += wgt
      if result[1]:
        morelist.append((scr, factor, result[1]))
    score = max(0, min(1, score / weight))
    more  = dict()
    if score < 1.0:
      # todo: this should be a slightly more intelligent selection process
      for scur, factor, reason in morelist:
        if scur < self.threshold:
          more[factor.category] = reason
    return (score, more)

#------------------------------------------------------------------------------
def test(value, extra=None):
  return Meter().test(value, extra=extra)

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------
